src/ClientData/TimerData.cpp

```diff
@@ -63,6 +65,43 @@ std::vector<const orbit_client_protos::TimerInfo*> TimerData::GetTimers(uint64_t
   return timers;
 }
 
+std::vector<const orbit_client_protos::TimerInfo*> TimerData::GetTimersAtDepthDiscretized(
+    uint32_t depth, uint32_t resolution, uint64_t start_ns, uint64_t end_ns) const {
+  ORBIT_SCOPE_WITH_COLOR("GetTimersAtDepthDiscretized", kOrbitColorBlueGrey);
+  absl::MutexLock lock(&mutex_);
+  // The query is for the interval [start_ns, end_ns], but it's easier to work with the close-open
+  // interval [start_ns, end_ns+1). We have to be careful with overflowing.
+  end_ns = std::max(end_ns, end_ns + 1);
+
+  if (timers_.find(depth) == timers_.end()) return {};
+
+  std::vector<const orbit_client_protos::TimerInfo*> discretized_timers;
+  uint64_t next_pixel_start_ns = start_ns;
+
+  // We are iterating through all blocks, but we don't have other option if we use TimerChain. A
+  // possible optimization could be not iterating after end_time but the complexity shouldn't
+  // change and this solution is simpler.
+  for (const auto& block : *timers_.at(depth)) {
+    // Several candidate timers might be in the same block.
+    while (block.Intersects(next_pixel_start_ns, end_ns)) {
+      const orbit_client_protos::TimerInfo* timer = block.LowerBound(next_pixel_start_ns);
+      if (timer == nullptr || timer->start() >= end_ns) break;
+      discretized_timers.push_back(timer);
+
+      // Use the time of next pixel boundary as a threshold to avoid returning several timers
+      // for the same pixel that will overlap after.
+      next_pixel_start_ns = GetNextPixelBoundaryTimeNs(timer->end(), resolution, start_ns, end_ns);
+
+      // If the timestamp of the next pixel is after the end, no more timers need to be added and we
+      // are returning early just because of performance.
+      if (next_pixel_start_ns >= end_ns) {
```

Comment:
I think here you would need to use the "original/including end_ns"
```suggestion
      if (next_pixel_start_ns > including_end_ns) {
```

Reply:
Replied before: https://github.com/google/orbit/pull/4147/files#r959417276